src/lib/BitcoinTxnParser.sol

```diff
@@ -0,0 +1,251 @@
+                }
+                break;
+            }
+        }
+
+        if (opReturnData.length == 0) revert NO_OP_RETURN_FOUND();
+        return opReturnData;
+    }
+
+    /// @notice Helper function to extract bytes from calldata
+    /// @param data Source calldata bytes
+    /// @param start Starting position in the source data
+    /// @param length Number of bytes to extract
+    /// @return result The extracted bytes
+    function extractBytes(bytes calldata data, uint256 start, uint256 length)
+        internal
+        pure
+        returns (bytes memory result)
+    {
+        result = new bytes(length);
+        for (uint256 i = 0; i < length; i++) {
+            result[i] = data[start + i];
+        }
+        return result;
+    }
+
+    /// @notice Parses a raw Bitcoin transaction into a structured format
+    /// @param rawTxn The raw Bitcoin transaction bytes
+    /// @return txn Structured transaction data containing version, outputs, and locktime
+    /// @dev Handles both legacy and segwit transaction formats
+    function parseTransaction(bytes calldata rawTxn) internal pure returns (Transaction memory txn) {
+        uint256 offset = 0;
+
+        // Parse version (4 bytes)
+        bytes memory versionBytes = extractBytes(rawTxn, offset, 4);
+        txn.version = uint32(bytes4(versionBytes));
+        offset += 4;
+
+        // Check for segwit flag
+        bool isSegwit = false;
+        if (rawTxn[offset] == 0x00 && rawTxn[offset + 1] == 0x01) {
+            isSegwit = true;
+            offset += 2;
+        }
+
+        // Parse input count (VarInt)
+        (uint256 inputCount, uint256 inputCountSize) = parseVarInt(rawTxn[offset:]);
+        offset += inputCountSize;
+
+        // Skip inputs
+        for (uint256 i = 0; i < inputCount; i++) {
+            offset += 36; // Previous output (32 + 4)
+
+            // Script length (VarInt)
+            (uint256 scriptLength, uint256 scriptLengthSize) = parseVarInt(rawTxn[offset:]);
+            offset += scriptLengthSize + scriptLength;
+
+            offset += 4; // Sequence
+        }
+
+        // Parse output count (VarInt)
+        (uint256 outputCount, uint256 outputCountSize) = parseVarInt(rawTxn[offset:]);
+        offset += outputCountSize;
+
+        // Parse outputs
+        txn.outputs = new Output[](outputCount);
+        for (uint256 i = 0; i < outputCount; i++) {
+            // Value (8 bytes)
+            bytes memory valueBytes = extractBytes(rawTxn, offset, 8);
+            txn.outputs[i].value = uint64(bytes8(valueBytes));
+            offset += 8;
+
+            // Script length (VarInt)
+            (uint256 scriptLength, uint256 scriptLengthSize) = parseVarInt(rawTxn[offset:]);
+            offset += scriptLengthSize;
+
+            // Script
+            txn.outputs[i].script = extractBytes(rawTxn, offset, scriptLength);
+            offset += scriptLength;
+        }
+
+        // Skip witness data
```

Comment:
Personally, i think we don't need to skip the witness data. This data will be helpful to validate the p2tr output script data.